File: apps/api/app/schemas/receipts.py
import uuid
from decimal import Decimal

from pydantic import BaseModel


class ReceiptSettingsOut(BaseModel):
    show_logo: bool
    show_customer_details: bool
    show_gst_breakdown: bool
    show_sku: bool
    show_cashier: bool
    show_qr_code: bool
    footer_message: str | None
    terms_and_conditions: str | None
    return_policy: str | None
    upi_id: str | None
    social_contact_info: str | None
    default_paper_width_mm: int

    class Config:
        from_attributes = True


class ReceiptSettingsUpdate(BaseModel):
    show_logo: bool | None = None
    show_customer_details: bool | None = None
    show_gst_breakdown: bool | None = None
    show_sku: bool | None = None
    show_cashier: bool | None = None
    show_qr_code: bool | None = None
    footer_message: str | None = None
    terms_and_conditions: str | None = None
    return_policy: str | None = None
    upi_id: str | None = None
    social_contact_info: str | None = None
    default_paper_width_mm: int | None = None


class ReceiptLineItem(BaseModel):
    name: str
    sku: str | None = None
    qty: Decimal
    uom: str
    rate: Decimal
    discount: Decimal = Decimal("0")
    tax_rate: Decimal = Decimal("0")
    line_total: Decimal


class ReceiptData(BaseModel):
    document_type: str
    document_number: str
    document_date: str
    document_time: str | None = None

    company_name: str
    company_legal_name: str
    company_gstin: str | None = None
    company_phone: str | None = None
    company_email: str | None = None
    company_address: str | None = None

    branch_name: str
    branch_gstin: str | None = None

    cashier_name: str | None = None

    customer_id: uuid.UUID | None = None
    customer_name: str | None = None
    customer_phone: str | None = None
    customer_gstin: str | None = None
    place_of_supply: str | None = None

    items: list[ReceiptLineItem]

    subtotal: Decimal
    total_discount: Decimal = Decimal("0")
    cgst_amount: Decimal = Decimal("0")
    sgst_amount: Decimal = Decimal("0")
    igst_amount: Decimal = Decimal("0")
    round_off: Decimal = Decimal("0")
    grand_total: Decimal

    payment_method: str | None = None
    amount_paid: Decimal | None = None
    balance_due: Decimal | None = None
    change_due: Decimal | None = None
    customer_credit_balance: Decimal | None = None

    notes: str | None = None
    settings: ReceiptSettingsOut

File: apps/web/src/lib/receipts.ts
import { useQuery } from "@tanstack/react-query";

import { apiFetch } from "@/lib/api";

export interface ReceiptSettings {
  show_logo: boolean;
  show_customer_details: boolean;
  show_gst_breakdown: boolean;
  show_sku: boolean;
  show_cashier: boolean;
  show_qr_code: boolean;
  footer_message: string | null;
  terms_and_conditions: string | null;
  return_policy: string | null;
  upi_id: string | null;
  social_contact_info: string | null;
  default_paper_width_mm: number;
}

export interface ReceiptLineItem {
  name: string;
  sku: string | null;
  qty: string;
  uom: string;
  rate: string;
  discount: string;
  tax_rate: string;
  line_total: string;
}

export type ReceiptDocumentType = "invoice" | "pos_receipt" | "payment_receipt" | "estimate" | "delivery_receipt" | "credit_note";

export interface ReceiptData {
  document_type: ReceiptDocumentType;
  document_number: string;
  document_date: string;
  document_time: string | null;

  company_name: string;
  company_legal_name: string;
  company_gstin: string | null;
  company_phone: string | null;
  company_email: string | null;
  company_address: string | null;

  branch_name: string;
  branch_gstin: string | null;

  cashier_name: string | null;

  customer_id: string | null;
  customer_name: string | null;
  customer_phone: string | null;
  customer_gstin: string | null;
  place_of_supply: string | null;

  items: ReceiptLineItem[];

  subtotal: string;
  total_discount: string;
  cgst_amount: string;
  sgst_amount: string;
  igst_amount: string;
  round_off: string;
  grand_total: string;

  payment_method: string | null;
  amount_paid: string | null;
  balance_due: string | null;
  change_due: string | null;
  customer_credit_balance: string | null;

  notes: string | null;
  settings: ReceiptSettings;
}

export const DOCUMENT_TYPE_LABELS: Record<ReceiptDocumentType, string> = {
  invoice: "Tax Invoice",
  pos_receipt: "Receipt",
  payment_receipt: "Payment Receipt",
  estimate: "Estimate",
  delivery_receipt: "Delivery Note",
  credit_note: "Credit Note",
};

export function useReceiptData(documentType: ReceiptDocumentType | null, documentId: string | null) {
  return useQuery({
    queryKey: ["receipt", documentType, documentId],
    queryFn: () => apiFetch<ReceiptData>(`/receipts/${documentType}/${documentId}`),
    enabled: !!documentType && !!documentId,
  });
}
